Explanation/development process

dviererbe opened the pull request:
> [!IMPORTANT]
> This PR is targeted against #53 for better readability of the diff. **Set the base to the `2.0-preview` branch before merging!**

This PR adds the initial content for the explanation article of the Ubuntu development process.

*(This PR is based on #54, fixed the spelling mistake in the branch name and rebased onto `explanation/merges-and-syncs`)*

dviererbe committed:
feat: add content for expl/development-process

dviererbe committed:
feat: Add placeholder how-to/request-freeze-exception

dviererbe committed:
docs: improve wording and fix spelling based on feedback

See #54

Co-authored-by: Sally Makin <sally.makin@canonical.com>
Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,320 @@
+
+Stabilisation and deadlines (freezes)
+-------------------------------------
+
+During development, caution is increasingly exercised in making changes
+to Ubuntu to ensure a stable state is reached in time for the final release
+date. The typical order and length of the various freezes can be seen on the
```
> It's not quite clear what the definition of a freeze is here. I think we could be a bit more explicit about what we mean by "freeze", to avoid using the term without making it really clear what the term means.

_Originally posted by @s-makin in https://github.com/canonical/ubuntu-packaging-guide/pull/54#discussion_r1624269513_

dviererbe commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,320 @@
+Stabilisation and deadlines (freezes)
+-------------------------------------
+
+During development, caution is increasingly exercised in making changes
+to Ubuntu to ensure a stable state is reached in time for the final release
+date. The typical order and length of the various freezes can be seen on the
+current Release Schedule, which is usually posted as a Discourse article under
+the `"Release" topic <https://discourse.ubuntu.com/c/release/>`_. 
+
+.. note::
+    In the past, the Release Schedule was published in the Ubuntu Wiki.
+    See for example the `release schedule of Ubuntu 20.04 LTS (FocalFossa) <https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule>`_.
+
+During freezes, exceptions must be requested to approve changes.
+See :doc:`how to request a freeze exception </how-to/request-freeze-exception>`.
```
@s-makin what do you think about:

```suggestion
During development, caution is increasingly exercised in making changes
to Ubuntu to ensure a stable state is reached in time for the final release
date. Modifications to the Ubuntu package archive get incrementaly restricted,
effectively freezing the state of the Ubuntu package archive. The
deadlines when these restrictions get enabled are called "freezes". During
freezes, exceptions must be requested to approve changes. See 
:doc:`how to request a freeze exception </how-to/request-freeze-exception>`.
The typical order and length of the various freezes can be seen on the
current Release Schedule, which is usually posted as a Discourse article under
the `"Release" topic <https://discourse.ubuntu.com/c/release/>`_. 

.. note::
    In the past, the Release Schedule was published in the Ubuntu Wiki.
    See for example the `release schedule of Ubuntu 20.04 LTS (Focal Fossa) <https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule>`_.
```

**Edit:** just noticed that there is a space missing in "FocalFossa"

s-makin replied:
I think this is great, much clearer!

dviererbe replied:
Thanks, committed with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/6425645543a860e8f087b99eeae60b1370c11540

dviererbe committed:
fix(typo): missing space in "FocalFossa"

Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe committed:
fix(typo): maintainer's -> maintainers

Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe committed:
ci(spell check): add words

- ABI as in "Application Binary Interface"
- deprioritised
- Numbat as in "Ubuntu 24.04 LTS (Noble Numbat)"
- roadmap
- Roadmap
- Tahr as in "Ubuntu 14.04 LTS (Trusty Tahr)"
- toolchain
- UIF as in "User Interface Freeze"

Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe committed:
feat: Improve wording for freeze exlanantion.

Incorporates feedback from Sally Makin <sally.makin@canonical.com>:
> It's not quite clear what the definition of a freeze is here.
> I think we could be a bit more explicit about what we mean by
> "freeze", to avoid using the term without making it really clear
> what the term means.
See:
- https://github.com/canonical/ubuntu-packaging-guide/pull/54#discussion_r1624269513
- https://github.com/canonical/ubuntu-packaging-guide/pull/62#discussion_r1668593715
- https://github.com/canonical/ubuntu-packaging-guide/pull/62#discussion_r1668615625

Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe committed:
fix(typo): incrementaly -> incrementally

Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe commented:
The failed CI checks are related to https://discourse.ubuntu.com/ and https://discourse.canonical.com/ being down. Nothing for me to action on, but the RTD build fail, which is weird.

EDIT: Failure is caused by:
`/home/docs/checkouts/readthedocs.org/user_builds/canonical-ubuntu-packaging-guide/checkouts/62/docs/explanation/development-process.rst:16: WARNING: term not in glossary: 'seeds'`

dviererbe committed:
feat(glossary): add term for Seeds

Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe commented:
Nice, all checks have passed here too now! 🎉

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+time-based release cycle, Ubuntu contributors must coordinate well to
+produce an on-time release.
+
+See also the article, :doc:`/explanation/releases`, for more details about the
```
```suggestion
See also the article :doc:`/explanation/releases` for more details about the
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+The broader open-source community gets together at the :term:`Ubuntu Summit` 
+(similar but different to the past 
+:term:`Ubuntu Developer Summits <Ubuntu Developer Summit>`) to share
+experiences and ideas and to inspire future exciting projects covering
```
```suggestion
experiences and ideas and to inspire future projects covering
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+The :term:`Canonical` teams meet after every release at the Roadmap Planning 
+Sprint to plan the roadmap for the next release and discuss how to implement
+the roadmap at the Engineering Sprint.
```
This is internal process info, and as such I think it shouldn't be included. While not a secret, it suggests there are parts of the devel process that community is excluded from (nothing wrong about the fact itself -- we just don't need to advertise it).
```suggestion
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+Sprint to plan the roadmap for the next release and discuss how to implement
+the roadmap at the Engineering Sprint.
+
+Merging with Upstream & feature development
```
```suggestion
Merging with upstream and feature development
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+The first phase of the release cycle is characterised by bringing new releases
+of :term:`upstream` components into Ubuntu, either directly or via 
+:doc:`Merges and Syncs from Debian </explanation/debian-merges-and-syncs>`. 
```
```suggestion
:doc:`Merges and syncs from Debian </explanation/debian-merges-and-syncs>`. 
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+The first phase of the release cycle is characterised by bringing new releases
+of :term:`upstream` components into Ubuntu, either directly or via 
+:doc:`Merges and Syncs from Debian </explanation/debian-merges-and-syncs>`. 
+Development of planned projects for the release often begins while merging is
```
```suggestion
The development of planned projects for the release often begins while merging is
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+of :term:`upstream` components into Ubuntu, either directly or via 
+:doc:`Merges and Syncs from Debian </explanation/debian-merges-and-syncs>`. 
+Development of planned projects for the release often begins while merging is
+still underway, and their development is accelerated once the package archive
```
```suggestion
still underway, and the development accelerates once the package archive
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+:ref:`DebianImportFreeze`.
+
+Stabilisation and deadlines (freezes)
+-------------------------------------
```
Unless it's a required term, I suggest using 'milestones' instead of the slightly morbid deadlines.
```suggestion
Stabilisation and milestones (freezes)
--------------------------------------
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+During development, caution is increasingly exercised in making changes
+to Ubuntu to ensure a stable state is reached in time for the final release
+date. Modifications to the Ubuntu package archive get incrementally restricted,
```
Can we specify _who_ does the restricting here? It would both make the information more complete and also allow us to avoid the use of the passive voice.

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+to Ubuntu to ensure a stable state is reached in time for the final release
+date. Modifications to the Ubuntu package archive get incrementally restricted,
+effectively freezing the state of the Ubuntu package archive. The
+deadlines when these restrictions get enabled are called "freezes". During
```
Same ^

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+date. Modifications to the Ubuntu package archive get incrementally restricted,
+effectively freezing the state of the Ubuntu package archive. The
+deadlines when these restrictions get enabled are called "freezes". During
+freezes, exceptions must be requested to approve changes. See 
```
```suggestion
freezes, developers must request exceptions to approve changes. See 
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+:doc:`how to request a freeze exception </how-to/request-freeze-exception>`.
+The typical order and length of the various freezes can be seen on the
+current Release Schedule, which is usually posted as a Discourse article under
+the `"Release" topic <https://discourse.ubuntu.com/c/release/>`_.
```
Just trying to reshuffle it to eliminate passive.
```suggestion
The release team usually posts the current Release Schedule as a Discourse article
under the `"Release" topic <https://discourse.ubuntu.com/c/release/>`_. It shows the typical order and length of the various freezes.
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+.. note::
+    In the past, the Release Schedule was published in the Ubuntu Wiki.
+    See for example the `release schedule of Ubuntu 20.04 LTS (Focal Fossa) <https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule>`_.
```
```suggestion
    See, for example, the `release schedule of Ubuntu 20.04 LTS (Focal Fossa) <https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule>`_.
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+Testing weeks
+~~~~~~~~~~~~~
+
+During a release's development phase, testing weeks are organised to focus the
```
```suggestion
During the release development phase, developers organise testing weeks to focus the
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
 ~~~~~~~~~~~~~~~~~~~~
 
+The automatic import of new packages and versions of existing packages from
+Debian gets disabled. The import of a new package or version of an existing
```
Again, who does the disabling, please?

dviererbe replied:
Archive admins disable the automatic import of changes from Debian

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
 ~~~~~~~~~~~~~~~~~~~
 
+At this point, Ubuntu developers should stop introducing new features,
+packages, and :term:`API`/:term:`ABI` changes, and instead concentrate on
```
Just curious: prior to this, there are no restrictions on ABI changes?

dviererbe replied:
No (based on the documentation I have read and the changes I have experienced).

Prior FF, it is effectively the wild west in the archive and also the reason why major disruptive archive event are announced in the release schedule (see e.g. https://discourse.ubuntu.com/t/oracular-oriole-release-schedule/36460)

If a change breaks reverse dependencies, this has to be addressed before the final release by patching the issue, rebuilding with the new dependencies or removing the reverse dependency (just a component or an entire package); also known as "proposed-migration" and "+1 Maintenance".

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+* the appearance of the desktop,
+* the distribution-specific artwork,
+* all user-visible strings in the desktop and applications that are installed 
+  by default.
```
```suggestion
* User interface of individual applications that are installed by default
* Appearance of the desktop
* Distribution-specific artwork
* All user-visible strings in the desktop and applications that are installed 
by default
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+Documentation strings should no longer be created or modified. This freeze
+ensures that the documentation can be accurately translated.
+
+Exceptions to this rule may be considered before release for significant and
```
```suggestion
Exceptions to this rule may be considered before the release for significant and
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+ensures that the documentation can be accurately translated.
+
+Exceptions to this rule may be considered before release for significant and
+glaring typographical errors or exceptional circumstances.
```
```suggestion
glaring errors or exceptional circumstances.
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+All new hardware enablement tasks for devices targeting the given release
+should be finished, and all the respective packages should be in the Ubuntu
+package archive. The release team will no longer accept changes in the Ubuntu
```
```suggestion
package archive. The release team no longer accepts changes in the Ubuntu
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+Beta Freeze
+~~~~~~~~~~~
+
+For the beta release's preparation, all uploads are queued and subject to
```
```suggestion
For the beta release preparation, all uploads are queued and subject to
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+For the beta release's preparation, all uploads are queued and subject to
+manual approval by the release team. Changes to packages that affect beta
+release images (flavours included) require the release team's approval before
+uploading. Uploads for packages that do not affect images will generally be
```
```suggestion
uploading. Uploads for packages that do not affect images are generally
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+Kernel Freeze
+~~~~~~~~~~~~~
+
+The Kernel Freeze is a deadline for kernel updates since they require several
```
```suggestion
The Kernel Freeze is the final date for kernel updates because they require several
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+.. _NonLanguagePackTranslationDeadline:
+
+Non-language-pack translation deadline
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
Non-language-pack translation final date
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

dviererbe replied:
issue: @rkratky this is the wording used in the release schedule (e.g. https://discourse.ubuntu.com/t/noble-numbat-release-schedule/35649), I think that it is confusing if we use a different name

dviererbe replied:
I will set this thread for now as resolved to unblock this PR from merging. In exchange I  opened #73.

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+they are subject to an earlier deadline to give time to developers to manually
+export translations from Launchpad and integrate them into the package.
+
+This deadline marks the date after which translations for such packages are not
```
```suggestion
This marks the date after which translations for such packages are not
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+^^^^^^^^^^^^^^^^^
+
+Packages in :ref:`ArchiveComponents_Universe` that aren't seeded in any of the
+Ubuntu flavours just remain in :ref:`FeatureFreeze` because they do not affect
```
```suggestion
Ubuntu flavours remain in :ref:`FeatureFreeze` because they do not affect
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+When the Final Release is close (~1.5 days out), developers should consider
+uploading to the :ref:`proposed pocket <ArchivePockets_Proposed>`, from which
+the release team will cherry-pick into the
```
```suggestion
the release team cherry-picks into the
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+the release team will cherry-pick into the
+:ref:`release pocket <ArchivePockets_Release>` if circumstances allow.
+All packages uploaded to the proposed pocket that do not make it into the
+release pocket until the Final Release will become candidates for
```
```suggestion
release pocket until the Final Release become candidates for
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+:ref:`StableReleaseUpdates_Summary`. Therefore, uploads to the proposed pocket
+during Final Freeze should meet the requirements of Stable Release Updates if
+the upload is not accepted into the release pocket. In particular, the upload
+must reference at least one bug, which will be used to track the stable update. 
```
```suggestion
must reference at least one bug, which is used to track the stable update. 
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+If you are sure that your upload will be accepted during Final Freeze, you can
+upload directly to the release pocket, but be aware that you have to re-upload
+after Final Release if the upload gets rejected.
```
```suggestion
.. note::
    If you are sure that your upload will be accepted during Final Freeze, you can
    upload directly to the release pocket, but be aware that you have to re-upload
    after Final Release if the upload gets rejected.
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+
+.. _LanguagePackTranslationDeadline:
+
+Language pack translation deadline
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
Language pack translation last day
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

dviererbe replied:
note: I agree that the word `deadline` should be avoided for obvious reasons, but it is the term that was and is used in Ubuntu release schedules (see for example the [Oracular Oriole Release Schedule](https://discourse.ubuntu.com/t/oracular-oriole-release-schedule/36460) or [Language Pack Translation Deadline (Ubuntu Wiki)](https://wiki.ubuntu.com/LanguagePackTranslationDeadline)).

thought: Maybe it is worth having a discussion with the release team to change the used terminology.

rkratky replied:
Thanks for the context. Please, consider the deadline-related suggestion non-blocking.

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
 Final Release
 -------------
+
+Once the :ref:`ReleaseCandidate` ISO is declared stable, it will be announced on the 
+`ubuntu-announce mailing list <https://lists.ubuntu.com/archives/ubuntu-announce/>`_
+and referred to as the "Final Release".
```
```suggestion
Once the release team declares the :ref:`ReleaseCandidate` ISO stable and names it
the "Final Release", a representative of the team announces it on the 
`ubuntu-announce mailing list <https://lists.ubuntu.com/archives/ubuntu-announce/>`_.
```

rkratky commented on docs/explanation/development-process.rst:
```diff
@@ -1,17 +1,322 @@
+and referred to as the "Final Release".
+
+.. note::
+    See for example the `Ubuntu 24.04 LTS (Noble Numbat) release announcement <https://lists.ubuntu.com/archives/ubuntu-announce/2024-April/000301.html>`_.
```
```suggestion
    See, for example, the `Ubuntu 24.04 LTS (Noble Numbat) release announcement <https://lists.ubuntu.com/archives/ubuntu-announce/2024-April/000301.html>`_.
```

rkratky reviewed (changes requested):
Thank you, @dviererbe, great stuff! I left a bunch of comments and suggestions.

About the passive voice: I realize it's not always possible or suitable, but in places where its use can be avoided, it's worth trying. The text is then easier to read, and sometimes it also forces us to provide more specific info instead of "hiding" behind the anonymity of the passive :)

dviererbe committed:
fix(expl/development-process): improve wording

Fixes grammar/spelling mistakes and improves the
used wording.

Co-authored-by: Robert Krátký <rkratky@gmail.com>
Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

dviererbe commented:
> About the passive voice: I realize it's not always possible or suitable, but in places where its use can be avoided, it's worth trying. The text is then easier to read, and sometimes it also forces us to provide more specific info instead of "hiding" behind the anonymity of the passive :)

I will keep this in mind for writing future texts :)
